src/flashcart/ed64/ed64_internal.c

```diff
@@ -0,0 +1,690 @@
+
+/* register functions (dependent on flashcart version) */
+
+/* register functions for V2 & V2.5  (These carts do not support FLASHRAM)*/
+/* The end of SDRAM is used for SRAM or FlashRAM save types */
+void ed64_bios_io_reg_v2(uint32_t address, uint32_t data) {
+
+    *(volatile uint32_t *) (ROM_ADDRESS);
+    *(volatile uint32_t *) (ROM_ADDRESS + address) = data;
+}
+
+/* register functions for V3 */
+void ed64_bios_io_reg_v3(uint32_t address, uint16_t data) {
+
+    ed64_bios_reg_write(REG_FLASHRAM_ADDR, address);
+    ed64_bios_reg_write(REG_FLASHRAM_DATA, data);
+}
+
+/* initialize functions (dependent on flashcart version) */
+
+/* Initilize V2 & 2.5 cart */
+void ed64_bios_init_v2() {
+
+    uint8_t buff[512];
+    uint16_t cfg = ed64_bios_reg_read(REG_CFG);
+
+    ed64_bios_reg_write(REG_CFG, 0);
+    ed64_bios_io_reg_v2(0xaa / 4 * 4, 0x00980098);
+    ed64_bios_dma_read_rom(buff, 0, 1);
+    ed64_bios_io_reg_v2(0xaa / 4 * 4, 0x00f000f0);
+    ed64_bios_dma_read_rom(buff, 0, 1);
+    ed64_bios_reg_write(REG_CFG, cfg);
+
+}
+
+/* Initilize V3 cart */
+void ed64_bios_init_v3() {
+
+    uint8_t buff[1024];
+    uint16_t cfg = ed64_bios_reg_read(REG_CFG);
+
+    ed64_bios_reg_write(REG_CFG, 0);
+    ed64_bios_reg_write(REG_CFG_COUNT, 161);
+    ed64_bios_io_reg_v3(0x55, 0x98);
+    ed64_bios_dma_read_rom(buff, 0, 2);
+    ed64_bios_io_reg_v3(0x55, 0xF0);
+    ed64_bios_dma_read_rom(buff, 0, 2);
+    ed64_bios_dma_read_rom(buff, 1024, 2);
+    ed64_bios_dma_read_rom(buff, 1024 + 256 - 2, 2);
+    ed64_bios_reg_write(REG_CFG_COUNT, 1);
+
+    ed64_bios_reg_write(REG_CFG, cfg);
+}
+
+/* Initialize cart */
+uint8_t ed64_bios_init() {
+
+    uint16_t firmware_ver;
+    uint16_t firmware_msg;
+    uint8_t cold_start;
+
+
+    IO_WRITE(PI_BSD_DOM2_LAT_REG, REG_LAT);
+    IO_WRITE(PI_BSD_DOM2_PWD_REG, REG_PWD);
+
+
+    ed64_bios_reg_write(REG_KEY, ED64_KEY_UNLOCK);
+    ed64_bios_reg_write(REG_CFG, 0x0000);
+
+
+    firmware_msg = ed64_bios_reg_read(REG_FPGA_FW_DATA);
+    cold_start = (firmware_msg & FPGA_FW_DATA_HOT_START) == 0 ? 1 : 0;
+    if (cold_start) {
+        firmware_msg |= FPGA_FW_DATA_HOT_START;
+        ed64_bios_reg_write(REG_FPGA_FW_DATA, firmware_msg);
+    }
+
+    firmware_ver = ed64_bios_reg_read(REG_FPGA_FW_VER);
+    if ((firmware_ver & 0xf000) >= 0x2000) {
+        ed64_bios_init_v3();
+    } else {
+        ed64_bios_init_v2();
+    }
+
+
+
+    spi_cfg = SPI_CFG_SS | BI_SPI_SPD_LO;
+    ed64_bios_reg_write(REG_CFG, ED_CFG_SDRAM_ON);
+    ed64_bios_reg_write(REG_SPI_CFG, spi_cfg);
+    ed64_bios_save_type = SAVE_TYPE_OFF;
+
+
+    return cold_start;
+}
+
+void ed64_bios_reset_spx() {
+
+    uint16_t cfg = ed64_bios_reg_read(REG_CFG);
+
+    ed64_bios_reg_write(REG_CFG, 0x8000);
+    wait_ms(100);
+    ed64_bios_reg_write(REG_CFG, cfg);
+    wait_ms(100);
+}
+
+uint32_t ed64_bios_reg_read(uint32_t reg) {
+
+    *(volatile uint32_t *) (ED64_REGS_BASE);
+    return *(volatile uint32_t *) (ED64_REGS_BASE + reg * 4);
```

Comment:
Is there something preventing us from using `io_read`/`io_write` functions provided by libdragon?

Reply:
I dont think so, but this is a bit out of my comfort zone and I kept the code similar to the original that I have with the thinking that it can be improved once we have something working.